File: cpp/MergeOverlappingIntervals.java
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class MergeOverlappingIntervals{
  public static void main(String[] args){
    List<int[]> intervals = generateRandomIntervals(10000, 0, 90);
    Solution s = new Solution();
    Instant startTime = Instant.now();
    System.out.println(s.mergeOverlappingIntervals(intervals));
    Instant endTime = Instant.now();
    Duration elapsedTime = Duration.between(startTime, endTime);
    long elapsedMilliseconds = elapsedTime.toMillis();
    System.out.println("Total Execution time in Java: " + elapsedMilliseconds);
  }

  public static List<int[]> generateRandomIntervals(int numIntervals, int minTime, int maxTime) {
    List<int[]> intervals = new ArrayList<>();
    Random random = new Random();

    for (int i = 0; i < numIntervals; i++) {
        int startTime = random.nextInt(maxTime - minTime + 1) + minTime;
        int endTime = random.nextInt(maxTime - startTime + 1) + startTime;

        intervals.add(new int[]{startTime, endTime});
    }

    return intervals;
  }
}

class Solution{
  public int mergeOverlappingIntervals(List<int[]> intervals){
    Collections.sort(intervals, (a, b) -> Integer.compare(a[1], b[1]));
    int overlappingIntervals = 0;
    int prevEnd = intervals.get(0)[1];
    for(int i = 1; i < intervals.size(); ++i){
      if(intervals.get(i)[0] >= prevEnd ) prevEnd = intervals.get(i)[0];
      else{
        ++overlappingIntervals;
        if(prevEnd < intervals.get(i)[1]) prevEnd = intervals.get(i)[1];
      }
    }
    return overlappingIntervals;
  }
}

File: cpp/merge_overlapping_intervals.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <random>

using namespace std;

class Solution{
  public:
    int eraseOverlappingIntervals(vector<vector<int>>& intervals){
      stable_sort(intervals.begin(), intervals.end(), [](const vector<int>& a, const vector<int>& b){
        return a[1] < b[1];
      });
      int overlappingIntervals = 0;
      int prevEnd = intervals[0][1];
      for(int i = 1; i < intervals.size(); ++i){
        if(intervals[i][0] >= prevEnd ) prevEnd = intervals[i][1];
        else{
          ++overlappingIntervals;
          if(prevEnd < intervals[i][i]) prevEnd = intervals[i][1];
        }
      }
      return overlappingIntervals;
    }
};

vector<vector<int>> generateRandomIntervals(int numIntervals, int minTime, int maxTime) {
    vector<vector<int>> intervals;
    std::mt19937 rng(std::random_device{}());
    std::uniform_int_distribution<> dist(minTime, maxTime);
    int minIntervalDuration = 10;
    for (int i = 0; i < numIntervals; ++i) {
        int startTime = dist(rng);
        int endTime = startTime + dist(rng);

        vector<int> interval = {startTime, endTime};
        intervals.emplace_back(interval);
    }

    return intervals;
}

int main(){
  Solution s;
  //vector<vector<int>> intervals = {{1,2}, {2,3}, {3,4}, {1,3}};
  vector<vector<int>> intervals = generateRandomIntervals(10000, 0, 90);
  auto start = chrono::high_resolution_clock::now();
  cout << s.eraseOverlappingIntervals(intervals) << endl;
  auto end = chrono::high_resolution_clock::now();
  auto elapsed_nanoseconds = chrono::duration_cast<chrono::nanoseconds>(end - start).count();
  auto elapsed_milliseconds = elapsed_nanoseconds / 1000000;
  cout << "execution time :" << elapsed_milliseconds << endl;
}